[SamuelFoo] iP

SamuelFoo opened the pull request:
# Duck

> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duck frees your mind of having to remember things you need to do. It's,

- text-based
- easy to learn
- ~~FAST~~ *SUPER* FAST to use

All you need to do is,

1. download it from [here](https://github.com/SamuelFoo/ip/).
1. double-click it.
1. add your tasks.
1. let it manage your tasks for you 😉

And it is **FREE**!

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)

If you are a Java programmer, you can use it to practice Java too. Here's the list of commands available represented as an `Enum`:


```java
public enum Command {
    LIST("list"),
    FIND("find"),
    MARK("mark"),
    UNMARK("unmark"),
    DELETE("delete"),
    BYE("bye"),
    TODO("todo"),
    DEADLINE("deadline"),
    EVENT("event");
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

SamuelFoo committed:
Rename "Duke" as "Duck"

SamuelFoo committed:
Add simple greeting.

SamuelFoo committed:
Add simple greeting

SamuelFoo committed:
Merge branch 'master' of https://github.com/SamuelFoo/ip

SamuelFoo committed:
Modify greeting

SamuelFoo committed:
Make the chatbot capable of echoing user input

SamuelFoo committed:
Add the ability to store whatever text entered by the user and display them back to the user when requested

SamuelFoo committed:
Rename "List" class as "TaskList"

SamuelFoo committed:
Add a `Task` class to represent tasks

SamuelFoo committed:
Replace Scanner with new Reader class

SamuelFoo committed:
Add the ability to mark tasks as done

SamuelFoo committed:
Add the ability to change the status back to not done

SamuelFoo committed:
Added "enums"

SamuelFoo committed:
Add different task types

SamuelFoo committed:
Add String placeholders for deadline and event tasks

SamuelFoo committed:
Update README.md

SamuelFoo committed:
Refactor

SamuelFoo committed:
Reorg

SamuelFoo committed:
Remove unused runtest.bat

SamuelFoo committed:
Semi-automate testing with multiple input files

SamuelFoo committed:
Fix null pointer exception when running test cases

SamuelFoo committed:
Add usage exceptions

SamuelFoo committed:
Add TODOs

SamuelFoo committed:
Add support for deleting tasks from the list

SamuelFoo committed:
Add the capability to save and load tasks from the disk

SamuelFoo committed:
Merge branch 'branch-Level-7'

SamuelFoo committed:
Store deadline dates as a java.time.LocalDate in the task objects

SamuelFoo committed:
Merge branch 'branch-Level-8'

SamuelFoo committed:
Rename CSVHandler as Storage

SamuelFoo committed:
Rename CommandBuffer as Parser

SamuelFoo committed:
Change storage file path to be relative to project directory

SamuelFoo committed:
Merge branch 'branch-A-MoreOOP'

SamuelFoo committed:
Merge remote-tracking branch 'upstream/add-gradle-support'

SamuelFoo committed:
Change main class name for gradle

SamuelFoo committed:
Add unit tests for deadline and todo

SamuelFoo committed:
Merge branch 'branch-A-JUnit'

SamuelFoo committed:
Fix bug where tasks data file is not updated after marking or unmarking tasks

SamuelFoo committed:
Update TODO

SamuelFoo committed:
Add documentation

SamuelFoo committed:
Remove unneeded Reader class

SamuelFoo committed:
Merge branch 'branch-A-JavaDoc'

SamuelFoo committed:
Add checkstyle to gradle

SamuelFoo committed:
Sort imports and remove trailing whitespaces

SamuelFoo committed:
Merge branch 'branch-A-CodingStandard'

SamuelFoo committed:
Use checkstyle config from addressbook

SamuelFoo committed:
Add changes to follow coding standards

SamuelFoo committed:
Give users a way to find a task by searching for a keyword in the task description

SamuelFoo committed:
Merge branch 'branch-Level-9'

adipanda2002 commented on src/main/java/Duck.java:
```diff
@@ -0,0 +1,161 @@
+                    // Tasks
+                    // TODO: Reduce duplicate code
+
+                    String taskPart = lineBuffer.getRemainingLine();
```
Do you think the name taskPart could cause some confusion regarding what 'part' is being referred to? Maybe something like taskDetails might be more relevant in this context.

adipanda2002 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @return tokens until before the pattern joined into a single line
+     */
+    public String getUntil(String pattern) {
```
What am I getting, and until when? I think the name getUntil could be replaced with something clearer or more descriptive.

adipanda2002 commented on src/test/java/tasks/DeadlineTest.java:
```diff
@@ -0,0 +1,31 @@
+    @Test
+    public void createDeadline_unmarked_success() throws Exception {
+        // toString
+        Deadline testDeadline1 = new Deadline("test description", new DateAndTime("2024-08-30"));
```
Could you replace these names that use numbers with more descriptive names that elaborate what type of edge case is being tested?

adipanda2002 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+    private final ArrayList<Task> tasks = new ArrayList<>();
+    // TODO: Refactor this
+    // Path relative to project directory
+    private static final String[] dirPathParts = { "store" };
```
Do you agree that using an abbreviation in dirPathParts might be a bit unclear?

adipanda2002 commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @return tokens until before the pattern joined into a single line
+     */
+    public String getUntilAndRemovePattern(String pattern) {
```
Would you agree that there's a lot of words in this name, but a lack of clarity as to what the method is doing?

adipanda2002 reviewed (commented):
LGTM. Generally, your naming is done quite well. There's a few nits with regards to naming here and there that I've commented on that would make the overall clarity of your code even better. Given how well you have used OOP, you could support that with optimizing your naming across the codebase.

elliot-tang commented on src/main/java/Duck.java:
```diff
@@ -0,0 +1,161 @@
+import java.util.Scanner;
+
+import exceptions.DeadlineUsageException;
+import exceptions.EventUsageException;
```
great job on organising your exception and task classes into packages! to adhere to the coding standard, perhaps you can consider adding all your other classes to a new package (e.g. the "duck" package) as well?

![image](https://github.com/user-attachments/assets/f2b70d02-1a6a-4a9e-bc09-e9dc782332a6)
attached the relevant portion of the coding standards for your reference.

elliot-tang commented on src/main/java/Duck.java:
```diff
@@ -0,0 +1,161 @@
+    private static final String GOODBYE = "Bye. Hope to see you again soon!";
+
+    // List to store user inputs
+    private static final TaskList TASKS = new TaskList();
```
great job ensuring that all of your constants have been declared with screaming snake case!

elliot-tang commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,138 @@
+    private final File saveDir = new File(String.join(File.separator, dirPathParts));
+    private static final Storage CSV_HANDLER = new Storage(
+            new File(String.join(File.separator, filePathParts)));
+
```
well done ensuring that your method name stays attached to the open parenthesis that follows it!

elliot-tang commented on src/main/java/Parser.java:
```diff
@@ -0,0 +1,122 @@
+     *
+     * @return true if the buffer has more tokens,
+     *         false otherwise
+     */
```
perhaps you can consider capitalising the first word and adding a full stop after your return descriptions, in order to format it like a sentence.

![image](https://github.com/user-attachments/assets/de173d2a-d9dc-4125-b10a-89d5ce65f164)
attached relevant portion of coding standards, see the last bullet point

elliot-tang commented on src/main/java/Duck.java:
```diff
@@ -0,0 +1,161 @@
+            Parser lineBuffer = new Parser(line);
+
+            // "bye" and "list" will only work if they are the only word in the line
+            // e.g. "bye bye" would not work
```
great job on ensuring proper indentation for both code and comments and following egyptian style brackets consistently throughout your code!

elliot-tang reviewed (commented):
LGTM! just a few nits to fix, but overall great job on adhering to the coding stadards for the course!

SamuelFoo committed:
Add Java FX to gradle build file

SamuelFoo committed:
Add barebones gui

SamuelFoo committed:
Update GUI styling

SamuelFoo committed:
Add response for GUI

SamuelFoo committed:
Re-organise classes into packages

SamuelFoo committed:
Add error handling for wrong datetime format

SamuelFoo committed:
Add greeting and tasks loading when starting up

SamuelFoo committed:
Enable assertions in gradle run

SamuelFoo committed:
Add indentation to tasks listing

SamuelFoo committed:
Add assertions when loading from storage

SamuelFoo committed:
Merge pull request #1 from SamuelFoo/branch-A-Assertions

Add assertions

SamuelFoo committed:
Refactor commands as classes

Code to handle the Todo, Deadline and Event
commands in the main class is extracted out into
individual command classes.

This improves code quality.

SamuelFoo committed:
Reduce levels of abstraction for path variables

SamuelFoo committed:
Refactor command classes

Code to handle the Delete, Mark and Unmark
commands in the main class is extracted out into
individual command classes.

SamuelFoo committed:
Merge pull request #2 from SamuelFoo/branch-A-CodeQuality

Refactor to improve the code quality

SamuelFoo committed:
Add do-after tasks

SamuelFoo committed:
Add screenshot of Ui

SamuelFoo committed:
Update TODO list

SamuelFoo committed:
Move all exception handling to command classes

SamuelFoo committed:
Change do-after task naming

SamuelFoo committed:
Change naming from time to date

SamuelFoo committed:
Update deadline naming

SamuelFoo committed:
Add user guide

SamuelFoo committed:
Add test for user guide quick-start commands

SamuelFoo committed:
Refactor command handling

SamuelFoo committed:
Update todo list

SamuelFoo committed:
Add Parser unit tests

SamuelFoo committed:
Edit display string when there are missing args

SamuelFoo committed:
Refactor TaskList

SamuelFoo committed:
Update todo list

SamuelFoo committed:
Handle bye for GUI

SamuelFoo committed:
Add exception handling for invalid taskList index

SamuelFoo committed:
Fix code quality issues

SamuelFoo committed:
Fix code quality issues

SamuelFoo committed:
Remove unneeded static variables

SamuelFoo committed:
Fix bug in testing code
